[deckhouse]: disruption updates (#2025)
deckhouse/deckhouse: modules/020-deckhouse/docs/internal/RELEASE.md

# DeckhouseRelease details

## Available annotations

| annotation                               | description                                                                        |
|------------------------------------------|------------------------------------------------------------------------------------|

> | release.deckhouse.io/force               | Apply specified release without any checks. Force deploy.                          |
> | release.deckhouse.io/disruption-approved | Approve release with disruptive changes. Works if update.disruptionMode is Manual. |
> | release.deckhouse.io/approved            | Approve release for deployment. Works if update.mode is Manual.                    |

Reviewer: ```suggestion
| `release.deckhouse.io/force`               | Apply specified release without any checks. Force deploy.                          |
| `release.deckhouse.io/disruption-approved` | Approve release with disruptive changes. Works if `update.disruptionMode` is Manual. |
| `release.deckhouse.io/approved`            | Approve release for deployment. Works if `update.mode` is Manual.                    |
```

## Disruptive release

It's a release with some potentially dangerous changes (change some default value / behavior / docker -> containerd, etc)
To handle this release, you should add disruption check logic in a release `X-1`, for example - register requirements.DisruptionFunc in init() [example](modules/402-ingress-nginx/hooks/requirements.go)
And then in a release `X` you should add record `"disruption:$functionName": "true"` to the [requirements.json](requirements.json) file.

> For example:

Reviewer: ```suggestion

For example:
```

In the release 1.35.0 set disruption check logic into ingressNginx

> In the release 1.36.0 set `"disruption:ingressNginx": "true"` to the requirements.json

Reviewer 2: Imagine stable release channel. Can we add a disruption check in 1.35.5 (not 1.35.0), and update to 1.36.10 for this machinery to work?

Author: We could and this will work but we want to avoid this behavior. Because in a some airgap installation client can jump from 1.35.3 to 1.36.10. And 1.35.3 wouldn't have this machinery, so we will meet a problem. It's just an agreement, cause as I said, machinery will work here

Reviewer 2: Ok. The development flow is clear to me:

To have a disruptive update in `1.X.*`, one should do:

1. Make a PR to `1.(X-1).0` with **both** precondition check and update requirement. In `requirements.json' the check should be turned on, but the update requirement is not.
2. Make PR  to `1.X.0` which turns the update requirement on in `requirements.json`.

> In the release 1.36.0 set `"disruption:ingressNginx": "true"` to the requirements.json

Reviewer 2: Can I add both check and disruption to 1.36.0? What's the point in splitting check and disruption between different versions?

Author: You could have
```
"ingressNginx": "1.1",
"disruption:ingressNginx": "true"
```
in the same release. It would work, cause these checks pass in a row

In my mind they are splitted like that:
 - RequirementCheck (like `"ingressNginx": "1.1"`, or `"k8s": "1.19"`) - it's some strict check of outer dependencies which deny a release deploy until a requirement is met. (**hard block**)
 - DisruptionCheck - it's a check, that warn user about some meaningful changes, which are controlled by our code and logic (simple: we are changing some behavior) (**soft block**)

Reviewer 2: So, in order to have Deckhouse in working state, we need to precaution clients in `X-1` release, because if the requirement is not met, release `X` will not work. The client will be notified, but Deckhouse will not be stuck because of the inevitable migration. I see.

## Release requirements

> For checking some precondition/requirement - register CheckFunc like [here](modules/402-ingress-nginx/hooks/requirements.go)
> and then add `"$functionName": "$version"` to the [requirements.json](requirements.json) file. (Example: `"ingressNginx": "0.33"`)

Reviewer: ```suggestion
For checking some precondition/requirement — register CheckFunc like [here](modules/402-ingress-nginx/hooks/requirements.go)
and then add `"$functionName": "$version"` to the [requirements.json](requirements.json) file. (e.g., `"ingressNginx": "0.33"`)
```

> and then add `"$functionName": "$version"` to the [requirements.json](requirements.json) file. (Example: `"ingressNginx": "0.33"`)

Reviewer 2: ```suggestion
and then add `"$functionName": "$version"` to the [requirements.json](requirements.json) file. (Example: `"disruption:ingressNginx": "0.33"`)
```

Author: Well, we are putting all this values into openapi values. They will look and process much more complicated, that why I decided to highlight it with the prefix

Reviewer 2: Okay. Nesting is inconvenient. 

What I still miss is the asymmetry in notations. I got it why I am wrong here. It would be more clear to me, if we had more explicit notations. E.g.:

```json
{
  "check:ingressNginx": "1.1",
  "disruption:ingressNginx": "1.1"
}
```

What do you think?

Reviewer 2: Let's stick to the domain and terminology a little bit. 

I guess, what we fill in `requirements.json` is not a disruption in any sense. It is a precondition check and an update requirement. We describe only the condition data and the fact of the requirement should be met:

1. precondition **check** is "1.1" for key "ingressNginx"
2. update **requirement** must be met (being "true") for "ingressNginx" to update Deckhouse

And there can be no disruption! 

- If the cluster has already had no default controller version, the check condition has already passed, and the requirement is already met. So there will be no disruption. 
- If a client notices the disruptive alert/notification and satisfies the **requirement**, there will be no disruption. 

So, I would pick more concise wording for this file.

```go
type ... interface {
    SetCoditionCheckFunc(string, func)
    Check(string)

    SetRequirementFunc(string, func)
    IsRequired(string)
}
```

```json
{
    "check:ingressNginx": "1.1",
    "require:ingressNginx": "true" // BTW why not boolean here?
}
```

What do you think?

# DeckhouseRelease details

## Available annotations

| annotation                               | description                                                                        |
|------------------------------------------|------------------------------------------------------------------------------------|
| `release.deckhouse.io/force`               | Apply specified release without any checks. Force deploy.                          |
| `release.deckhouse.io/disruption-approved` | Approve release with disruptive changes. Works if `update.disruptionApproval` is Manual. |
| `release.deckhouse.io/approved`            | Approve release for deployment. Works if `update.mode` is Manual.                    |

## Difference between disruption check and requirement check

> All release settings, requirements and disruptions are stored in the file [release.yaml](release.yaml)

Reviewer: ```suggestion
All release settings, requirements, and disruptions are stored in the [release.yaml](release.yaml) file.
```

> - RequirementCheck (like `"ingressNginx": "1.1"`, or `"k8s": "1.19"` in the section `requirements`) - it's some strict check of outer dependencies which deny a release deploy until a requirement is met. (hard block)
> - DisruptionCheck - (like `"ingressNginx"` in the section `disruptions`) - it's a check, that warn user about some meaningful changes, which are controlled by our code and logic (simple: we are changing some behavior / default value) (soft block)

Reviewer: ```suggestion
- RequirementCheck (like `"ingressNginx": "1.1"`, or `"k8s": "1.19"` in the section `requirements`) — it's some strict check of outer dependencies which deny a release deploy until a requirement is met (hard block).
- DisruptionCheck - (like `"ingressNginx"` in the section `disruptions`) — it's a check that warn a user about some meaningful changes, which are controlled by our code and logic (simple: we are changing some behavior or default value) (soft block).
```

### Disruptive release

> It's a release with some potentially dangerous changes (change some default value / behavior / docker -> containerd, etc)

Reviewer: ```suggestion
It's a release with some potentially dangerous changes (e.g. which changes some default value, behavior, or changes docker to containerd, etc.).
```

To handle this release, you should add disruption check logic in a release at least `X-1` (previous release), for example — register DisruptionFunc in init() [example](modules/402-ingress-nginx/hooks/requirements.go).
And add record for a specified release, where this logic will be checked. (e.g.: `"1.36": ["ingressNginx"]` to the [release.yaml](release.yaml) file, section `disruptions`.

> Example:

Reviewer 2: ```suggestion
How to add a disruptive change:
```

> - In current release (say, 1.35.0) set disruption check logic into the `ingressNginx` function;

Reviewer: ```suggestion
- In current release (say, 1.35.0) set disruption check logic into the `ingressNginx` function.
```
